connmgr/connmanager.go

```diff
@@ -0,0 +1,235 @@
+		log.Debugf("Retrying connection to %v in: %v", c.Addr, d)
+		time.AfterFunc(d, func() {
+			cm.Connect(c)
+		})
+	} else if uint32(len(cm.conns)) < cm.cfg.MaxOutboundConnections {
+		go cm.Connect(&ConnReq{Addr: cm.cfg.GetNewAddress()})
+	}
+}
+
+// connHandler handles all connection related requests.  It must be run as a
+// goroutine.
+//
+// The connection handler makes sure that we maintain a pool of active outbound
+// connections so that we remain connected to the network.  Connection requests
+// are processed and mapped to their corresponding network connections.
+func (cm *ConnManager) connHandler() {
+out:
+	for {
+		select {
+		case req := <-cm.requests:
+			switch msg := req.(type) {
+
+			case handleConnected:
+				msg.c.state = ConnEstablished
+				cm.conns[msg.c] = msg.conn
+				log.Infof("Connection established: %v", msg.c.Addr)
+				msg.c.retryCount = 0
+
+				if cm.cfg.OnConnection != nil {
+					go cm.cfg.OnConnection(msg.c, msg.conn)
+				}
+
+			case handleDisconnected:
+				msg.c.state = ConnDisconnected
+				if conn, ok := cm.conns[msg.c]; ok {
+					conn.Close()
+				}
+				delete(cm.conns, msg.c)
+				log.Debugf("Connection closed: %v", msg.c.Addr)
+
+				cm.handleFailedConn(msg.c)
+
+			case handleFailed:
+				msg.c.state = ConnFailed
+				log.Debugf("Connection failed %v: %v", msg.c.Addr, msg.err)
+
+				cm.handleFailedConn(msg.c)
+			}
+
+		case <-cm.quit:
+			break out
+		}
+	}
+
+	cm.wg.Done()
+	log.Trace("Connection handler done")
+}
+
+// Connect dials a connection to the requested address.
+// It can be run as a goroutine so as to not block during the connection
+// attempt.
+func (cm *ConnManager) Connect(c *ConnReq) {
+	log.Debugf("Attempting connection: %v", c.Addr)
+	conn, err := cm.cfg.Dial("tcp", c.Addr)
+	if err != nil {
+		cm.requests <- handleFailed{c, err}
+	} else {
+		cm.requests <- handleConnected{c, conn}
+	}
+}
+
+// Disconnect disconnects the connection corresponding to the given connection
+// request.
+func (cm *ConnManager) Disconnect(c *ConnReq) {
+	log.Debugf("Disconnecting %v", c.Addr)
+	cm.requests <- handleDisconnected{c}
+}
+
+// Start launches the connection manager and begins connecting to the network.
+func (cm *ConnManager) Start() {
+	log.Trace("Connection manager started")
+	cm.wg.Add(1)
+	go cm.connHandler()
+
+	for i := uint32(0); i < cm.cfg.MaxOutboundConnections; i++ {
+		go cm.Connect(&ConnReq{Addr: cm.cfg.GetNewAddress()})
+	}
+}
+
+// Wait blocks until the connection manager halts gracefully.
+func (cm *ConnManager) Wait() {
+	cm.wg.Wait()
+}
+
+// Stop gracefully shuts down the connection manager.
+func (cm *ConnManager) Stop() {
+	if atomic.AddInt32(&cm.stop, 1) != 1 {
+		log.Warnf("Connection manager already stopped")
+		return
+	}
+	close(cm.quit)
+	log.Trace("Connection manager stopped")
+}
+
+// New returns a new connection manager.
+// Use Start to start connecting to the network.
+func New(cfg *Config) (*ConnManager, error) {
+	cm := ConnManager{
+		cfg:      cfg,
```

Comment:
I'd suggest making the `ConnManager` field a concrete type and making this a copy like `peer` does [here](https://github.com/btcsuite/btcd/blob/master/peer/peer.go#L2062).  This will prevent the caller from being able to mutate it after the fact.

So:
```Go
		cfg:      *cfg,
```